lib/rules/require-var-decl-first.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * Requires `var` declaration to be on the top of an enclosing scope
+ *
+ * Types: `Boolean`
+ *
+ * Values: `true`
+ *
+ * if `requireVarDeclFirst` defined as a `true` value, it will report errors of `var` declarations that
+ * does not appear on the top of a function scope.
+ *
+ * #### Example
+ *
+ * ```js
+ * "requireVarDeclFirst": true
+ * ```
+ *
+ * ##### Valid
+ *
+ * ```js
+ * var x = 1,
+ *     y = 2;
+ * ```
+ * ```js
+ * var x = 1;
+ * var y = 2;
+ * ```
+ * ```js
+ * var x = 1;
+ * // comments
+ * var y = 2;
+ * ```
+ * ```js
+ * var x = 1;
+ * // comments
+ * // comments 2
+ * var y = 2;
+ * ```
+ * ```js
+ * const a = 1;
+ * const b = 2;
+ * ```
+ * ```js
+ * var x = 1;
+ * function y() {var z;};
+ * ```
+ * ```js
+ * var x = 1;
+ * var y = function () {var z;};
+ * ```
+ * ```js
+ * var w = 1;
+ * function x() {
+ *  var y;
+ * // comments
+ * // comments 2
+ *  var z;
+ * };
+ * ```
+ * ```js
+ * var x = 1;
+ * var y;
+ * for (y = 0; y < 10; y++) {};
+ * ```
+ *
+ * ##### Invalid
+ *
+ * ```js
+ * var x;
+ * x = 1;
+ * var y = 2;
+ * ```
+ * ```js
+ * var w = 1;
+ * function x() {var y;};
+ * var z = 2;
+ * ```
+ * ```js
+ * var w = 1;
+ * function x() {
+ *  var y;
+ *  y = 2;
+ *  var z;
+ * };
+ * ```
+ * ```js
+ * var a;
+ * for(var count=0;count < 10;count++){}
+ * ```
+  * ```js
+ * var x;
+ * for(var count=0;count < 10;count++){
+ *  var y;
+ * }
+ * ```
+ *
+ */
+
+var assert = require('assert');
+
+module.exports = function() {};
+
+module.exports.prototype = {
+    configure: function(options) {
+        assert(
+            options === true,
+            this.getOptionName() + ' option requires a true value'
+        );
+    },
+
+    getOptionName: function() {
+        return 'requireVarDeclFirst';
+    },
+
+    check: function(file, errors) {
+        var parentNodes = {};
+        var commentTokens = [];
+
+        file.iterateTokensByType(['Line', 'Block'], function(token) {
+            commentTokens.push(token);
+        });
+
+        file.iterateNodesByType(['VariableDeclaration'], function(node) {
+            var count;
+            var comment;
+            var commentLength;
+            var parentNode;
+            var parentNodeStack;
+            var lastNode;
+            var whiteSpaceLength;
+
+            parentNode = node.parentNode;
+            while (parentNode.type !== 'Program' &&
```

Comment:
Ah, i see. Youre trying to find the wrapping function of the node,  if any. Yes, a function like getFunctionContaining(node) would help clarify.